icelake/src/io/task_writer.rs

```diff
@@ -147,3 +169,216 @@ impl UnpartitionedWriter {
         self.data_file_writer.close().await
     }
 }
+
+/// Partition task writer
+pub struct PartitionedWriter {
+    schema: ArrowSchemaRef,
+    operator: Operator,
+    table_location: String,
+    location_generator: Arc<DataFileLocationGenerator>,
+    partition_spec: PartitionSpec,
+    partition_type: Any,
+    table_config: TableConfigRef,
+    /// # TODO
+    ///
+    /// Support to config the data file writer.
+    groups: HashMap<OwnedRow, PartitionGroup>,
+}
+
+struct PartitionGroup {
+    pub partition_array: StructArray,
+    pub writer: DataFileWriter,
+}
+
+impl PartitionedWriter {
+    /// Create a new `PartitionedWriter`.
+    pub fn new(
+        schema: ArrowSchemaRef,
+        table_location: String,
+        location_generator: DataFileLocationGenerator,
+        partition_spec: PartitionSpec,
+        partition_type: Any,
+        operator: Operator,
+        table_config: TableConfigRef,
+    ) -> Self {
+        Self {
+            schema,
+            operator,
+            table_location,
+            location_generator: location_generator.into(),
+            partition_spec,
+            partition_type,
+            table_config,
+            groups: HashMap::new(),
+        }
+    }
+
+    /// This function do two things:
+    /// 1. Partition the batch by partition spec.
+    /// 2. Create the partition group if not exist.
+    ///
+    /// # TODO
+    /// Mix up two thing in one function may be not a good idea.
+    /// The reason do that is we need to use the partition info in step 1 when we create the
+    /// partition group. To avoid data copy, we do that in one function. We may need to refactor
+    /// this function in the future. Also the name of this function may not be a good choice.
+    async fn split_by_partition(
+        &mut self,
+        batch: &RecordBatch,
+    ) -> Result<HashMap<OwnedRow, RecordBatch>> {
+        let value_array = Arc::new(StructArray::from(
+            self.partition_spec
+                .fields
+                .iter()
+                .map(|field| {
+                    let array: ArrayRef = batch
+                        .column_by_name(&field.name)
+                        .ok_or_else(|| {
+                            crate::error::Error::new(
+                                crate::ErrorKind::IcebergDataInvalid,
+                                format!(
+                                    "Can't find column according partition spec, column name {}",
+                                    field.name
+                                ),
+                            )
+                        })?
+                        .clone();
+                    let transform_func = create_transform_function(&field.transform)?;
```

Comment:
Why not keep this function in writer?

Reply:
This function is used to create TransforFunction from transform. From this general functionality, place it in transform crate may more appropriate. Or as a method of Transform.🤔

Reply:
No, I mean we don't need to create a function for every batch? We just need to create it once for a partitoned writer?

Reply:
Get it! Yes we should do it once.